Updates Result Summaries

trevorb1 opened the pull request:
<!--- Provide a short description of the changes in the Title -->

### Description
<!--- Describe your changes in detail -->
Migrates from the the old results summary to new format; does not change the logic. Replaces the old node discount cost with updated result from otoole. 

~~**NOTE**: We can not merge this until otoole 1.1.4 is released.~~

### Issue Ticket Number
<!--- Link corresponding issue number -->
na

### Documentation
<!--- Where and how has this change been documented -->
na

maartenbrinkerink commented on resources/otoole.yaml:
```diff
@@ -142,7 +142,7 @@ OperationalLifeStorage:
     type: param
     dtype: float
-    default: 10
+    default: 1
```
Was this change deliberate? Doesn't matter either way as OL is defined within the workflow was just wondering

trevorb1 replied:
Was deliberate, and I just aligned it with powerplant operational life default values

maartenbrinkerink commented on resources/otoole.yaml:
```diff
@@ -42,7 +42,7 @@ CapitalCostStorage:
     type: param
     dtype: float
-    default: 2000
+    default: 0.001
```
Was this change deliberate? Doesn't matter either way as storage costs are defined within the workflow was just wondering

trevorb1 replied:
Was deliberate, and I just aligned it with powerplant capital cost default values

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/constants.py:
```diff
@@ -0,0 +1,35 @@
+"""Constants for summary statistics"""
+
+RENEWABLES = ["GEO", "HYD", "SPV", "CSP", "WAS", "WAV", "WON", "WOF"]
```
BIO should probable be included in RENEWABLES here as this is currently also the case in other constant files (e.g. powerplant/constants.py).

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/gen_shares.py:
```diff
@@ -0,0 +1,78 @@
+from constants import CLEAN, RENEWABLES, FOSSIL
+
+# do not include storage in generation share calcualtion 
+EXCLUDE = ["LDS", "SDS"]
```
This should be read dynamically from the storage_parameters entry in the config file since the storage technologies can change.

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/headline.py:
```diff
@@ -0,0 +1,121 @@
+
+    return df[
+        (df.index.get_level_values("TECHNOLOGY").str.startswith("PWR"))
+        & ~(df.index.get_level_values("TECHNOLOGY").str.startswith("PWRLDS"))
```
This should be read dynamically from the storage_parameters entry in the config file since the storage technologies can change.

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/summarise_results.py:
I'd link this to the constants file as well as such that the renewables_filter/fossil_filter are alligned with the entries from constants.

trevorb1 replied:
I agree! I moved this to a new issue though, so we can do this all at once for all subdirectories. We will need to create `__init__.py` files for each directory and change import locations. So for future ref, it will be nice to have a single PR that does all these changes across many files.

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/trade_flows.py:
Perhaps on Monday you can walk me through how the values in TradeFlows.csv are derived. Takes me a bit too much time right now to figure that out:)

maartenbrinkerink replied:
Can we also add domestic and international transmission flows as separate entries? E.g. `NODE_1` and `NODE_2` could be `CANXX`, 'USAXX' and 'CAN', 'CAN' for domestic flows. Note that for domestic flows negative values should first be turned positive and summed after.

trevorb1 replied:
Totally! This remained unchanged from how they were calculated before (ie. it used @abhishek0208's implementation)

maartenbrinkerink commented on workflow/snakefile:
```diff
@@ -31,7 +32,6 @@ def get_otoole_data(otoole_config: str, var: str) -> list[str]:
         "StorageLevelYearStart",
         "StorageLevelYearFinish",
         "Trade",
-        "DiscountedOperatingCost",
```
This removal caused the workflow on my end to not run (I think). I've added a commit to revert this, if that's incorrect let me know.

trevorb1 replied:
Good catch!

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/gen_shares.py:
Can we also add gen_shares at the country level?

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/node_cost.py:
Can we also add costs at the country level (simple sum)?

trevorb1 replied:
Add both total costs and power generation costs at country/node levels!

maartenbrinkerink commented on workflow/scripts/osemosys_global/summary/capacity.py:
Can we also add capacities at the country level (simple sum by tech)?

maartenbrinkerink replied:
Can we also add domestic and international transmission capacities as separate entries? E.g. `FROM` and `TO` could be `CANXX`, 'USAXX' and 'CAN', 'CAN' for domestic capacities. In the current setup international transmission capacities will already be reported at nodal level, but there are often multiple nodes that import/export from/to the same international node (e.g. multiple nodes between CAN & USA).

trevorb1 replied:
I believe both of these points have been solved in this commit now! I will do a double check that the international trading is working correctly in the morning with a more robust example.

maartenbrinkerink commented:
@trevorb1 This looks good! Two fixed should be made first I think (changing storage exclusions dynamically and making sure that RENEWABLE filters are in line) with some other suggestions provided (e.g. country level results).

trevorb1 commented on workflow/scripts/osemosys_global/summary/carbon_intensity.py:
Good call! Emission intensity was using production, but it should probably be using demand. I have updated the logic in this commit, but a confirmation that demand is correct would be great! Thanks! :)

maartenbrinkerink replied:
I think in Slack we said that it should be the total production - production from storage and transmission. I.e. this way it would also cover the extra generation required to compensate for storage and transmission losses which would not be taken into account when using demand.

trevorb1 replied:
Corrected in [this commit ](https://github.com/OSeMOSYS/osemosys_global/pull/219/commits/03249193e988c49156c67420fb20a3c4b74e12c6)

trevorb1 commented:
Thanks for the detailed feedback, @maartenbrinkerink! I believe most of your comments have now been addressed! One note, I (sorta) reverted your change of adding `DiscountedOperatingCost` to the missing otoole files list. Ended up being a typo as it should be `Operational` instead of `Operating`. Corrected in [this commit](https://github.com/OSeMOSYS/osemosys_global/pull/219/commits/fc2a0d17b022746caea98e12f0b91d0bf1cfb5d8).

maartenbrinkerink commented:
@trevorb1 I can't run the workflow at the moment. `MissingOutputException` occurs on my end for a numer of parameters (see below). 

![image](https://github.com/user-attachments/assets/04715bbd-0a84-4eeb-936a-2aaa0be01d9f)

For the below this can be solved by adding them to `missing` in the snakefile (not sure if that is the correct approach) but when trying to apply this to `"DiscountedCostByStorage"` the run still fails as the parameter is called in the `calculate_costs` rule but doesn't exist. I see that there is a `try/except FileExistsError` in `node_costs.py` but perhaps this is not functioning as it should?
```

"AnnualTechnologyEmissionsPenalty",
"CapitalInvestmentStorage",
"DiscountedCapitalInvestmentStorage",
"DiscountedSalvageValueStorage",  
```

trevorb1 commented:
Thanks for flagging those errors, @maartenbrinkerink! I have:
- Updated the emission intensity to use production with transmission and storage to be removed (pulling storage names from the config file) 
- Corrected issues resulting in failed workflows if storage/emission penalties are not included. This involved add the files you mentioned to the list of missing otoole files, with the exception of `AnnualTechnologyEmissionsPenalty`. This isnt a supported calculation (as only discounted emission penalty values are supported). That was my fault for not catching it when I merged otoole config file updates for `v1.1.4`. I have removed this from the otoole config file.
